_posts/18-05-2022-a_story_unfolds.md

```diff
@@ -0,0 +1,223 @@
+```haskell
+type LookAroundM m a = a -> m [a]
+type BuildNodeM t m a = a -> [t a] -> m (t a)
+
+treeBuilderM :: Monad m => LookAroundM m a -> BuildNodeM t m a -> a -> m (t a)
+treeBuilderM lookAroundM buildNodeM value = lookAroundM value >>= childNodes >>= buildNodeM value
+  where childNodes = traverse (treeBuilderM lookAroundM buildNodeM)
+```
+Now it contains the scary M-word, but it lets us redefine `treeBuilder` as a one-liner:
+```haskell
+treeBuilder lookAround buildNode value = runIdentity (treeBuilderM (return . lookAround) (\a b -> return (buildNode a b)) value)
+```
+The `lookAround` function now looks like this:
+```haskell
+lookAround :: LookAroundM IO File
+lookAround file = do
+  isDirectory <- isDir file
+  if isDirectory
+  then listDir file
+  else return []
+```
+and the `buildNode` becomes:
+```haskell
+buildNode :: File -> [FileTree File] -> IO (FileTree File)
+buildNode file subtrees = do
+  isDirectory <- isDir file
+  if isDirectory
+  then return (Directory file subtrees)
+  else return (RegularFile file)
+```
+with a minor modification to our `FileTree`:
+```haskell
+data FileTree a = RegularFile a | Directory a [FileTree a]
+```
+and to the `printTree` type signature:
+```haskell
+printTree :: FileTree File -> String
+```
+With all of this in place, we can rewrite our `buildTree` function:
+```haskell
+buildTree :: File -> IO (FileTree File)
+buildTree = treeBuilderM lookAround buildNode
+```
+Wonderfully concise.
+
+And thus, after a long walk among the trees, we arrive at a blissfully abstract, much less tangible function that traverses _any_ tree-like structure - as long as we are able to look around and see how it branches out. But is this useful in any way, or was it just an exercise in axe-swinging?
+
+The benefit of separating listing and printing is quite obvious, because these are in fact completely unrelated tasks - but, to reiterate, it gives us more flexibility in how we want to do any of them (change the formatting without worrying about introducing listing-related bugs), more code reusability (as now we can use the filesystem traversal to, for example, create or delete files along certain paths), ease of testing (as we can test these in isolation) and reasoning (as there is less cognitive overhead when reading code that does _one_ thing instead of _two_ things tangled together). The wonderful nature of functions lets us compose the building blocks back together however we please, to either recover the original intent, or to introduce new functionality.
+
+What we get from separating the tree traversal from listing new branches is, again, quite subtle. First, we know that our traversal works for every structure that at least resembles a tree - and it is ensured by Haskell type system. Second, finding stop condition-related bugs is now easier, as we shifted the responsibility for terminating the recursion to functions that _don't have to be recursive at all!_ Imagine how easy testing becomes!
```

Comment:
> Second, finding stop condition-related bugs is now easier, as we shifted the responsibility for terminating the recursion to functions that _don't have to be recursive at all!_

Could you elaborate on that point? Esp. by bringing up the names of the functions in question

Reply:
I'll come up with some examples.